Content.Server/Ninja/Systems/BatteryDrainerSystem.cs

```diff
@@ -0,0 +1,100 @@
+using Content.Server.Power.Components;
+using Content.Server.Power.EntitySystems;
+using Content.Shared.DoAfter;
+using Content.Shared.Interaction;
+using Content.Shared.Ninja.Components;
+using Content.Shared.Ninja.Systems;
+using Content.Shared.Popups;
+using Robust.Shared.Audio;
+
+namespace Content.Server.Ninja.Systems;
+
+/// <summary>
+/// Handles the doafter and power transfer when draining.
+/// </summary>
+public sealed class BatteryDrainerSystem : SharedBatteryDrainerSystem
+{
+    [Dependency] private readonly BatterySystem _battery = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedDoAfterSystem _doAfter = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<BatteryDrainerComponent, BeforeInteractHandEvent>(OnBeforeInteractHand);
+    }
+
+    /// <summary>
+    /// Start do after for draining a power source.
+    /// Can't predict PNBC existing so only done on server.
+    /// </summary>
+    private void OnBeforeInteractHand(EntityUid uid, BatteryDrainerComponent comp, BeforeInteractHandEvent args)
+    {
+        var target = args.Target;
+        if (args.Handled || comp.BatteryUid == null || !HasComp<PowerNetworkBatteryComponent>(target))
+            return;
+
+        // handles even if battery is full so you can actually see the poup
+        args.Handled = true;
+
+        if (_battery.IsFull(comp.BatteryUid.Value))
+        {
+            _popup.PopupEntity(Loc.GetString("battery-drainer-full"), uid, uid, PopupType.Medium);
+            return;
+        }
+
+        var doAfterArgs = new DoAfterArgs(uid, comp.DrainTime, new DrainDoAfterEvent(), target: target, eventTarget: uid)
+        {
+            BreakOnUserMove = true,
+            MovementThreshold = 0.5f,
+            CancelDuplicate = false,
+            AttemptFrequency = AttemptFrequency.StartAndEnd
+        };
+
+        _doAfter.TryStartDoAfter(doAfterArgs);
+    }
```

Comment:
Shouldn't this just be InteractHand.

Reply:
interacthand is raised on the target not the user